EdFi.Ods.Utilities.Migration.Tests/PgSql/MigrationTests/Latest/PostgresGlendaleTests.cs

```diff
@@ -0,0 +1,203 @@
+        protected override DatabaseRestoreSetupOption DatabaseRestoreSetupOption { get; } = DatabaseRestoreSetupOption.RestoreDuringFixtureSetupOnly;
+        protected override string TestDataDirectoryName => "Latest";
+
+        protected OdsUpgradeResult UpgradeResult { get; private set; }
+
+        protected override string OptionalTestSourceOdsBackupFullPath => EnsureTestGlenDaleBackupPathExists();
+
+        protected override string PsqlExecutable => EnsurePsqlExecutablePathExists();
+
+        protected override string TestDisabledReason => string.IsNullOrEmpty(GetFullGlenDaleBackupPath())
+            ? "GlenDale tests are currently optional due to disk space/runtime requirements. A GlenDale backup file was not supplied, so this test will be automatically disabled.  To enable, specify a valid backup file path in the application config"
+            : string.Empty;
+
+        protected string GetGlenDaleBackupDownloadUrl()
+        {
+            var grandBendBackupPathsByVersion = PostgreSqlMigrationTestSettingsProvider.GetGlenDaleBackupPaths();
+            if (grandBendBackupPathsByVersion.ContainsKey(FromVersion))
+            {
+                return grandBendBackupPathsByVersion[FromVersion];
+            }
+
+            throw new ApplicationException($"No GlenDale backup copy Path found for version {FromVersion.DisplayName}.");
+        }
+
+        private static string GetFullGlenDaleBackupPath()
+        {
+            return string.IsNullOrEmpty(PostgreSqlMigrationTestSettingsProvider.GetConfigVariable("GlendaleBackupPath"))
+                ? null
+                : Path.GetFullPath(PostgreSqlMigrationTestSettingsProvider.GetConfigVariable("GlendaleBackupPath"));
+        }
+        
+        protected string EnsureTestGlenDaleBackupPathExists()
+        {
+            var GlenDaleBackupDownloadUrl = GetGlenDaleBackupDownloadUrl();
+            var GlenDaleBackupsDirectory = Path.Combine(TestContext.CurrentContext.TestDirectory, "GlenDaleBackups");
+
+            if (!Directory.Exists(GlenDaleBackupsDirectory))
+            {
+                Directory.CreateDirectory(GlenDaleBackupsDirectory);
+            }
+
+            var tempDirectory = Path.Combine(GlenDaleBackupsDirectory, FromVersion.Value.ToString());
+            var versionSpecificBackupFileName = $"EdFi_Ods_GlenDale_{FromVersion.Value}_PG11.sql";
+            var GlenDaleBackupFilePath = Path.Combine(tempDirectory, versionSpecificBackupFileName);
+            if (File.Exists(GlenDaleBackupFilePath))
+            {
+                return GlenDaleBackupFilePath;
+            }
+
+            if (Directory.Exists(tempDirectory))
+            {
+                Directory.Delete(tempDirectory);
+            }
+
+            Directory.CreateDirectory(tempDirectory);
+
+            File.Copy(GlenDaleBackupDownloadUrl, GlenDaleBackupFilePath, true);
+
+            return GlenDaleBackupFilePath;
+        }
+        private string EnsurePsqlExecutablePathExists()
```

Comment:
Why are we duplicating this code?

Reply:
I refactored the code . Please review again